### ISSUE TITLE
Error thrown when setting dynamic current and vehicle not charging, expected?

### ISSUE THREAD

**Issue by jacoscar** [nordicopen/easee_hass#455]
### The problem

I've created an automation that uses the service "set circuit dynamic limit" for my Easee One charger.

I always get this error when executing it with the car is not charging, is this to be expected?

### Version of Easee integration having the issue?

0.9.59

### Version of Home Assistant Core having the issue?

2024.9.0

### Anything in the logs that might be useful for us?

```txt
Logger: homeassistant.components.automation.limit_dynamic_current
Source: helpers/script.py:525
integration: Automation (documentation, issues)
First occurred: 10:04:46 AM (1 occurrences)
Last logged: 10:04:46 AM

Limit dynamic current: Error executing script. Unexpected error for call_service at pos 3: 'dynamicCircuitCurrentP1'
Traceback (most recent call last):
  File "/config/custom_components/easee/controller.py", line 760, in check_circuit_current
    or charger_data.state[compare_p2] != current_p2
       ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pyeasee/utils.py", line 55, in __getitem__
    if type(self._storage[key]) == str and validate_iso8601(self._storage[key]):
            ~~~~~~~~~~~~~^^^^^
KeyError: 'dynamicCircuitCurrentP2'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 525, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 764, in _async_call_service_step
    response_data = await self._async_run_long_action(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 727, in _async_run_long_action
    return await long_task
           ^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2763, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2806, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/easee/services.py", line 582, in circuit_execute_set_current
    circuit = controller.check_circuit_current(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/easee/controller.py", line 766, in check_circuit_current
    charger_data.config[compare_p1] != current_p1
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pyeasee/utils.py", line 55, in __getitem__
    if type(self._storage[key]) == str and validate_iso8601(self._storage[key]):
            ~~~~~~~~~~~~~^^^^^
KeyError: 'dynamicCircuitCurrentP1
```

**Comment by olalid** [comment]
I think this and your other problem is one and the same really.
Did this start happening after upgrading HA to the latest version possibly?

**Comment by olalid** [comment]
Probably same root cause as https://github.com/nordicopen/easee_hass/issues/454

**Comment by jacoscar** [comment]
> I think this and your other problem is one and the same really. Did this start happening after upgrading HA to the latest version possibly?

I don't know, as when I installed the integration for the first time I was asked to upgrade HA, so I ended up on 2024.9.0. My previous HA version was not supported.

**Comment by olalid** [comment]
Ok, I have not myself tried it with 2024.9.0 yet, so I will try that later this evening and see if I can reproduce the errors you are seeing.

**Comment by olalid** [comment]
No, it works for me with 2024.9.0...
If you have not, can you please enable logging for Easee (see readme) and then search for any rows containing state_dynamicCircuitCurrentP1, state_dynamicCircuitCurrentP2 or state_dynamicCircuitCurrentP3 in the log file after doing a restart?
You should see some lines like this, indicating where values for the circuit current limits are received from the cloud.

´´´
2024-09-05 17:31:51.527 DEBUG (MainThread) [custom_components.easee.controller] Observation state_dynamicCircuitCurrentP1 0 type 3 <class 'int'>
2024-09-05 17:31:57.616 DEBUG (MainThread) [custom_components.easee.controller] Callback EXXXX 111 state_dynamicCircuitCurrentP1 0.0 3
2024-09-05 17:31:51.527 DEBUG (MainThread) [custom_components.easee.controller] Observation state_dynamicCircuitCurrentP2 0 type 3 <class 'int'>
2024-09-05 17:31:57.616 DEBUG (MainThread) [custom_components.easee.controller] Callback EXXXXX 112 state_dynamicCircuitCurrentP2 0.0 3
2024-09-05 17:31:51.527 DEBUG (MainThread) [custom_components.easee.controller] Observation state_dynamicCircuitCurrentP3 5 type 3 <class 'int'>
2024-09-05 17:31:57.616 DEBUG (MainThread) [custom_components.easee.controller] Callback EXXXX 113 state_dynamicCircuitCurrentP3 5.0 3
´´´

**Comment by olalid** [comment]
Hang on. Easee One is a one phase version, right?
It could be that it provides the data in a different way...
Do you have anything that says Callback EXXXXX 112 state_dynamicCircuitCurrentP1/2/3 at all?
Do you have any log file rows that says Unsupported data id?

**Comment by jacoscar** [comment]
> Hang on. Easee One is a one phase version, right? It could be that it provides the data in a different way... Do you have anything that says Callback EXXXXX 112 state_dynamicCircuitCurrentP1/2/3 at all? Do you have any log file rows that says Unsupported data id?


````2024-09-05 18:32:26.100 DEBUG (MainThread) [custom_components.easee.controller] Observation state_circuitTotalAllocatedPhaseConductorCurrentL1 6 type 3 <class 'int'>
2024-09-05 18:32:26.100 DEBUG (MainThread) [custom_components.easee.controller] Observation state_circuitTotalPhaseConductorCurrentL1 0.01899999938905239 type 3 <class 'float'>
2024-09-05 18:32:26.100 DEBUG (MainThread) [custom_components.easee.controller] Observation state_chargerFirmware 58 type 4 <class 'int'>
2024-09-05 18:32:26.100 DEBUG (MainThread) [custom_components.easee.controller] Observation state_reasonForNoCurrent 79 type 4 <class 'int'>
2024-09-05 18:32:26.100 DEBUG (MainThread) [custom_components.easee.controller] Observation state_smartCharging False type 2 <class 'bool'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_cableLocked True type 2 <class 'bool'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_cableRating 32 type 4 <class 'int'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_chargerOpMode 4 type 4 <class 'int'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_outputPhase 0 type 4 <class 'int'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_dynamicCircuitCurrentP1 6 type 3 <class 'int'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_outputCurrent 6 type 3 <class 'int'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_deratingActive False type 2 <class 'bool'>
2024-09-05 18:32:26.101 DEBUG (MainThread) [custom_components.easee.controller] Observation state_errorString Initial clock sync issues: num retries=1, fallback to non-tls=0 type 6 <class 'str'>
````
> Hang on. Easee One is a one phase version, right? It could be that it provides the data in a different way... Do you have anything that says Callback EXXXXX 112 state_dynamicCircuitCurrentP1/2/3 at all? Do you have any log file rows that says Unsupported data id?


````

2024-09-05 18:32:27.037 DEBUG (MainThread) [custom_components.easee.controller] Callback XXXXXXXXX 11 state_chargerOfflineReason 0 4
2024-09-05 18:32:27.038 DEBUG (MainThread) [custom_components.easee.controller] Callback XXXXXXXXX 15 config_localPreAuthorizeEnabled True 2
2024-09-05 18:32:27.038 DEBUG (MainThread) [custom_components.easee.controller] Callback XXXXXXXXX 16 config_localAuthorizeOfflineEnabled True 2
2024-09-05 18:32:27.038 DEBUG (MainThread) [custom_components.easee.controller] Callback XXXXXXXXX 17 config_allowOfflineTxForUnknownId True 2
2024-09-05 18:32:27.038 DEBUG (MainThread) [custom_components.easee.controller] Callback XXXXXXXXX 21 config_detectedPowerGridType 3 4
2024-09-05 18:32:27.039 DEBUG (MainThread) [custom_components.easee.controller] Callback XXXXXXXXX 22 config_circuitMaxCurrentP1 40.0 3
2024-09-05 18:32:27.039 DEBUG (MainThread) [custom_components.easee.controller] Callback XXXXXXXXX 30 state_lockCablePermanently False 2````





> No, it works for me with 2024.9.0... If you have not, can you please enable logging for Easee (see readme) and then search for any rows containing state_dynamicCircuitCurrentP1, state_dynamicCircuitCurrentP2 or state_dynamicCircuitCurrentP3 in the log file after doing a restart? You should see some lines like this, indicating where values for the circuit current limits are received from the cloud.
> 
> ```
> 2024-09-05 17:31:51.527 DEBUG (MainThread) [custom_components.easee.controller] Observation state_dynamicCircuitCurrentP1 0 type 3 <class 'int'>
> 2024-09-05 17:31:57.616 DEBUG (MainThread) [custom_components.easee.controller] Callback EXXXX 111 state_dynamicCircuitCurrentP1 0.0 3
> 2024-09-05 17:31:51.527 DEBUG (MainThread) [custom_components.easee.controller] Observation state_dynamicCircuitCurrentP2 0 type 3 <class 'int'>
> 2024-09-05 17:31:57.616 DEBUG (MainThread) [custom_components.easee.controller] Callback EXXXXX 112 state_dynamicCircuitCurrentP2 0.0 3
> 2024-09-05 17:31:51.527 DEBUG (MainThread) [custom_components.easee.controller] Observation state_dynamicCircuitCurrentP3 5 type 3 <class 'int'>
> 2024-09-05 17:31:57.616 DEBUG (MainThread) [custom_components.easee.controller] Callback EXXXX 113 state_dynamicCircuitCurrentP3 5.0 3
> ```

**Comment by olalid** [comment]
Ok, so it reports only the state_dynamicCircuitCurrentP1 etc...

Strange that your original post log file extract says
KeyError: 'dynamicCircuitCurrentP1 in that case, that specific key should be present.
I would expect it to fail with KeyError: 'dynamicCircuitCurrentP2' and KeyError: 'dynamicCircuitCurrentP3', but not with KeyError: 'dynamicCircuitCurrentP1', so not sure I understand why that happens...

In any case, I guess we need to change some things to be able to support these one phase versions...

**Comment by olalid** [comment]
@jacoscar if you are able, please try this PR https://github.com/nordicopen/easee_hass/pull/456

**Comment by olalid** [comment]
@jacoscar I suspect you have a similar issue as this one also?
https://github.com/nordicopen/easee_hass/issues/451
Can you please tell us which product code the Easee One has? Should be visible in same way in the GUI as in that issue.

**Comment by jacoscar** [comment]
> @jacoscar I suspect you have a similar issue as this one also? #451 Can you please tell us which product code the Easee One has? Should be visible in same way in the GUI as in that issue.

![image](https://github.com/user-attachments/assets/c1500991-68e2-4a1e-913c-d2155d274378)

**Comment by jacoscar** [comment]
> @jacoscar if you are able, please try this PR #456

![image](https://github.com/user-attachments/assets/92620ea0-79f1-4131-959d-31ff86404c11)

it works and I don't see the error when changing the dynamic circuit current limit anymore

thanks!

**Comment by olalid** [comment]
Thanks for testing, please let us know if you encounter anything else later.

**Comment by olalid** [comment]
Yes, you would need to enter at template that uses the current value of the setting and add or subtract from that to get the new value. Which should be possible to do if you switch to YAML mode.

**Comment by olalid** [comment]
This issue is most likely solved in v0.9.60. Please update and test.

**Comment by jacoscar** [comment]
tested and ok
